Objects/codeobject.c

```diff
@@ -2020,8 +2039,26 @@ _PyCode_Hydrate(PyCodeObject *code)
             return NULL;
         }
     }
-    Py_INCREF(pyc->names);
-    code->co_names = pyc->names;  // The items may still be NULL!!!
+
+    if (code->co_strings_start) {
+        assert(0); // this should not be active yet
+        // Why can't we just do:
+        // code->co_names = pyc->names + code->co_strings_start
+        // ?
+        if (code->co_strings_size) {
+            code->co_names = PyTuple_New(code->co_strings_size);
+            if (code->co_names == NULL) {
+                return NULL;
+            }
+        }
```

Comment:
I tried to follow your outline here: https://github.com/faster-cpython/ideas/issues/32#issuecomment-868744438
but I'm not sure why we are allocating a co_names array on the code object. Should we not be pointing into the module level one (which is bigger now because it has entries for all the duplicates, with redirects)?

Reply:
co_names is used by LOAD_GLOBAL, LOAD_ATTR and many others. My plan is to switch from a single big shared one to one per code object, since for those instructions the oparg is used as the index into co_names. Same as for co_consts, really.

Now, we *could* avoid the individual co_consts and co_names tuples and link to the shared one, and just add the local index bias to the index into the shared tuple. But then we'd have to change the Python API to only return a slice of the tuple (in addition to fully hydrating the values), and really, we'd do well to rename these members to something else (since anyone else using these will be in for a surprise when they don't know to add the index bias).

Reply:
One thing we're running into here is that traditionally all the data structures hanging off code objects are Python objects. That was always easy for memory management, but you can't have a pointer into the middle of a tuple object. In the longer term we may need to reconsider that.

Reply:
Yes, I see - if we go for a globals shared array we would need the python API to give a view into it or something.

Reply:
In my dream version all the data hanging off a code object is just pointers into the big PYC array, and the Python-level attributes are properties that construct new bytes objects in the fly (or a memoryview, but I think bytes would be fine -- we wouldn't even need to cache them).

This could be very efficient, since the interpreter usually just gets a pointer to the first data byte using PyBytes_AsString() or similar. (The only exceptioni s probably the names of keyword args, although I haven't looked at that mechanism in ages.)

But that would require updating the compiler. So I think for the prototype we have to stay compatible with the data structures as returned by the compiler, and pay the price. But we can put hydrating certain things off as long as possible -- as an example, if a function has something like
```
if unlikely_error_condition(value):
    raise ValueError("Some long error message")
```
if we're just a little lucky we'll never have to hydrate that error string.